R/geocode.R

```diff
@@ -0,0 +1,101 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#' Geocode a Location Using Various Methods
+#'
+#' @description This function takes a location name as a string and returns the latitude and longitude coordinates using various geocoding methods.
+#'   It supports multiple geocoding services and handles API keys where necessary.
+#'
+#' @usage geocode_raw(location, method = "osm")
+#'
+#' @param location A character string specifying the location name to geocode. It must be provided as a single string.
+#' @param method A character string specifying the geocoding method to use. Defaults to "osm" (OpenStreetMap). Supported methods include:
+#'   \code{"osm"}, \code{"census"}, \code{"arcgis"}, \code{"census_simple"}, \code{"geocodio"}, \code{"mapbox"},
+#'   \code{"google"}, \code{"bing"}, \code{"here"}, \code{"tomtom"}, \code{"nominatim"}, and \code{"tiger"}.
+#'
+#' @details This function attempts to geocode the specified location name using the chosen method. If the method requires an API key,
+#'   it will check for the necessary environment variable and return an error if the key is missing. The function outputs a data frame
+#'   containing the geocoded coordinates and location information.
+#'
+#' @return A data frame containing the following columns:
+#'   \item{name}{The name of the city or location.}
+#'   \item{country}{The country in which the location is situated, inferred from the display name.}
+#'   \item{lat}{The latitude of the location.}
+#'   \item{long}{The longitude of the location.}
+#'   \item{population}{The population of the location (not available in this function, returns \code{NA}).}
+#'
+#' @examples
+#' \dontrun{
+#'   # Geocode a location using OpenStreetMap (default)
+#'   geocode_raw("New York, USA")
+#'
+#'   # Geocode a location using Google (API key required)
+#'   geocode_raw("Tokyo, Japan", method = "google")
+#' }
+#'
+#' @export
+geocode <- function(location, method = "osm") {
+  if (missing(location) || !is.character(location)) {
+    stop("Please provide a valid location name as a string.")
+  }
+  
+  # Validate geocoding method
+  supported_methods <- c("osm", "census", "arcgis", "census_simple", "geocodio",
+                         "mapbox", "google", "bing", "here", "tomtom", "nominatim", "tiger")
+  if (!(method %in% supported_methods)) {
+    stop("Unsupported geocoding method. Please choose one of: ", paste(supported_methods, collapse = ", "))
+  }
+  
+  # Handle API keys for methods that require them
+  methods_with_keys <- c("google", "bing", "here", "tomtom", "mapbox", "geocodio")
+  if (method %in% methods_with_keys) {
+    api_key_env <- paste0(toupper(method), "_API_KEY")
+    api_key <- Sys.getenv(api_key_env)
+    if (api_key == "") {
+      stop(paste0("API key for ", method, " is required. Please set the '", api_key_env, "' environment variable."))
+    }
+  }
+  
+  result <- tryCatch({
+    geocode_df <- tibble::tibble(address = location) %>%
```

Comment:
`tidygeocoder::geocode(.tbl = data.frame(address=location), "address", method = method, quiet = TRUE)` gets rid of both the dependency on `tibble` and `dplyr`, which is nice.

Reply:
I could use the `.geocode `function as a fallback if no location found in the database but then how to let the user choose the geocoding service ? Should me prompt him/her and also provide an explanation in the README about the API keys ? Or do we systematically use the default free `osm `geocoder ?